### PR TITLE
#20. The solution for Valid Parentheses problem.

### PR DESCRIPTION
**Pull request by daniilnaumovich**

### CODE REVIEW

File: ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                {'(', ')'},
+                {'{', '}'},
+                {'[', ']'}
+            };
+
+            for (int i = 0; i < s.Length; i++)
+            {
+                if (parentheses.ContainsKey(s[i]))
+                {
+                    st.Push(s[i]);
```
Comment (shomzascnsoft) [inline ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs:20]: please use continue to get rid last else block;

File: ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                {
+                    st.Pop();
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            if (st.TryPeek(out char c))
```
Comment (shomzascnsoft) [inline ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs]: as.Any()

File: ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                {'[', ']'}
+            };
+
+            for (int i = 0; i < s.Length; i++)
+            {
+                if (parentheses.ContainsKey(s[i]))
+                {
+                    st.Push(s[i]);
+                }
+                else if (st.TryPeek(out char p) && parentheses[p] == s[i])
```
Comment (shomzascnsoft) [inline ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs]: if (st.Any() && parentheses[st.Pop()] == s[i]) it seems more preferable.

File: ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs [hunk 4]
```diff
@@ -18,18 +18,18 @@ public static bool IsValid(string s)
...
                     st.Push(s[i]);
+                    continue;
                 }
-                else if (st.TryPeek(out char p) && parentheses[p] == s[i])
-                {
-                    st.Pop();
-                }
-                else
+
+                if (st.Any() && parentheses[st.Pop()] == s[i])
```
Comment (shomzascnsoft) [inline ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs]: I'm not sure, but st.TryPop() can be the best option here.

**Review by shomzascnsoft — APPROVED** [review]

File: ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs (hunk 4 above)
Comment (daniilnaumovich) [reply ScnSoft-Education/LeetCodeTraining/ValidParentheses/ValidParenthesesSolution.cs]: hello @shomzascnsoft thank you for suggestion, this code works and it does indeed look like the most suitable option:
```
//if (st.Any() && parentheses[st.Pop()] == s[i])
if (st.TryPop(out char p) && parentheses[p] == s[i])
{
    continue;
}
```
I guess I should've used it from the very start, but for some reason I thought that TryPeek'ing and comparing the element before actually Popping it in this particular case was more correct. But apparently we can TryPop the element anyways and check the result, so I agree that TryPop works the best there.

**Pull request merged**